### ISSUE TITLE
Tabs widget incorrectly echos content in its run.

### ISSUE THREAD

**Issue by SamMousa** [yiisoft/yii2#7417]
https://github.com/yiisoft/yii2-bootstrap/blob/master/Tabs.php#L124

Instead of echo it should be returned.

**Comment by samdark** [comment]
Both are actually supported.

**Comment by SamMousa** [comment]
The docs in base\Widget say it should return a string. By allowing this random generation of output (Yii2 moved away from doing that, instead returning content) i have to use ob_start around every widget if I want to capture its output..

**Comment by SamMousa** [comment]
/**
     * Executes the widget.
     * @return string the result of widget execution to be outputted.
     */
    public function run()
    {
    }

**Comment by SamMousa** [comment]
Ty, also I double checked this happens in a lot of the widgets.

**Comment by samdark** [comment]
Well, yeah. You're welcome to make a pull request fixing it.

**Comment by SamMousa** [comment]
Done, note that I'm not sure what the right approach is for widgets that create output in their init as well as their run.

On the other hand, since init() is called by the constructor, I'm not sure it's good practice to create output when creating an object. (side effects and all that)

Not sure of a fix though.

--- Okay, will re-add the trailing \n.

**Comment by SamMousa** [comment]
http://www.yiiframework.com/doc-2.0/guide-structure-widgets.html

This guide amongst other things (like indeed saying direct echo is allowed) provides an example widget that uses ob_start in init() to capture everything it needs in the run(). That could be a viable approach.

Not sure if this still qualifies as a bug or more as an improvement / refactoring thing.

**Comment by samdark** [comment]
ob_start one is capturing widget.

**Comment by SamMousa** [comment]
Yes I know, but a widget that outputs tags in its init() function will need to capture output between init and run() if it wants to return the result instead of echoing it.

**Comment by samdark** [comment]
Well, yes.

**Comment by SamMousa** [comment]
So, my question then remains:
- Suppose we all agree that echoing stuff from the constructor (which calls init()) or from the run() function is not desired.

How are we (I'm up for making the changes if we agree on a solution) going to refactor the widgets like Modal that have content at 2 sides?

ob_start is the only way to reliable capture the output between init and run, so that is one way to go.
If we go that way though, in my opinion it would be cleaner to not echo anything in the init.

Basically:

        init() {
         // if this is a capturing widget.
        ob_start()
        }
        
        run() {
         return "stuff that used to be echoed in the init()" . ob_get_clean() . "stuf that used to be echoed in        the run()";
        }

**Comment by samdark** [comment]
That's OK to echo in init of there's `ob_start()`.

**Comment by SamMousa** [comment]
Fair enough, so do you want me to create another pull requests for all widgets that currently have init() functions with echo in the init and add ob_start to them?

**Comment by samdark** [comment]
What do you want to change in this case?

**Comment by SamMousa** [comment]
https://github.com/yiisoft/yii2-bootstrap/blob/master/NavBar.php

Widgets that use echo in their init() and run(). The changes will basically add ob_start() to the init function and run will then return ob_get_clean().

**Comment by samdark** [comment]
Ah, yes,
